src/blocks.js

```diff
@@ -1,40 +1,107 @@
+/* eslint-env serviceworker */
 import { S3Client } from '@aws-sdk/client-s3'
 import { DynamoDBClient } from '@aws-sdk/client-dynamodb'
 import { digest } from 'multiformats'
 import { CID } from 'multiformats/cid'
 import { fromString } from 'uint8arrays/from-string'
-import { DynamoIndex } from './s3/block-index.js'
+import { DynamoIndex, CachingIndex } from './s3/block-index.js'
 import { BatchingDynamoBlockstore } from './s3/blockstore.js'
+import { DenyingBlockStore } from './deny.js'
 
 const CAR = 0x202
 
 /**
+ * @typedef {import('multiformats').UnknownLink} UnknownLink
  * @typedef {import('./worker.js').Env} Env
-
+ *
  * @typedef {object} BlockLocation
  * @property {string} blockmultihash e.g zQmXWxE2sdWFeLHfVEp6bb6kib9YWWEvsaAJSgoExKrPVXD
  * @property {string} carpath e.g us-east-2/dotstorage-staging-0/raw/bafybeieiltf3tnfdyvdutyolzhfahphgevnjsso26nulfqxtkptyefq3za/315318734258473269/ciqjxmllx5y73brw6mv3pkvd7sotfk2turkupkq7tsgygrdy2yxibri.car
  * @property {number} length e.g 58 (bytes)
  * @property {number} offset e.g 96 (bytes)
+ *
+ * @typedef {object} Blockstore
+ * @prop {(cid: UnknownLink) => Promise<boolean>} has
+ * @prop {(cid: UnknownLink) => Promise<Uint8Array | undefined>} get
  */
 
 /**
  * @param {Env} env
+ * @param {ExecutionContext} ctx
  */
-export function getBlockstore (env) {
+export async function getBlockstore (env, ctx) {
   const dynamo = new DynamoIndex(getDynamoClient(env), env.DYNAMO_TABLE, { maxEntries: 3 })
-  const s3 = new BatchingDynamoBlockstore(dynamo, getS3Clients(env))
-  return new DagHausBlockStore(dynamo, env.CARPARK, s3)
+  const index = new CachingIndex(dynamo, await caches.open(`dynamo:${env.DYNAMO_TABLE}`), ctx)
+  const s3 = new BatchingDynamoBlockstore(index, getS3Clients(env))
+  const r2 = new DagHausBlockStore(index, env.CARPARK, s3)
```

Comment:
I'd have something like a `TieredBlockstore` that abstracts the fallback logic, like [this](https://github.com/web3-storage/content-claims/blob/main/packages/lambda/src/lib/store/tiered.js) for example, but not blocking...